src/QsCompiler/BondSchemas/CompilerDataStructures.bond

```diff
@@ -0,0 +1,911 @@
+}
+
+// Completely represents F#'s CharacteristicsKind.
+// N.B. Does not have an equivalent F# type.
+struct CharacteristicsKindComposition<T>
+{
+    5: required CharacteristicsKind Kind = InvalidSetExpr;
+
+    // Not null when Kind is SimpleSet.
+    10: nullable<OpProperty> SimpleSet = nothing;
+
+    // Not null when Kind is Union or Intersection.
+    15: nullable<CharacteristicsKindSetOperation<T>> SetOperation = nothing;
+}
+
+struct ResolvedCharacteristics
+{
+    5: required CharacteristicsKindComposition<ResolvedCharacteristics> Expression;
+}
+
+struct InferredCallableInformation
+{
+    5: required bool IsSelfAdjoint;
+
+    10: required bool IsIntrinsic;
+}
+
+struct CallableInformation
+{
+    5: required ResolvedCharacteristics Characteristics;
+
+    10: required InferredCallableInformation InferredInformation;
+}
+
+// N.B. Does not completely represent F#'s QsTypeKind.
+//      Used by QsTypeKindComposition to do a complete representation of the original F# type.
+enum QsTypeKind
+{
+    UnitType,
+    Int,
+    BigInt,
+    Double,
+    Bool,
+    String,
+    Qubit,
+    Result,
+    Pauli,
+    Range,
+    ArrayType,
+    TupleType,
+    UserDefinedType,
+    TypeParameter,
+    Operation,
+    Function,
+    MissingType,
+    InvalidType
+}
+
+// Used to encapsulate data when QsTypeKind is Function.
+// N.B. Does not have an equivalent F# type.
+struct QsTypeKindFunction<T>
+{
+    5: required T DataA;
+
+    10: required T DataB;
+}
+
+// Used to encapsulate data when QsTypeKind is Operation.
+// N.B. Does not have an equivalent F# type.
+struct QsTypeKindOperation<TData, TCharacteristics>
+{
+    5: required TData DataA;
+
+    10: required TData DataB;
+
+    15: required TCharacteristics Characteristics;
+}
+
+// Completely represents F#'s QsTypeKind.
+// N.B. Does not have an equivalent F# type.
+struct QsTypeKindComposition<TData, TUdt, TParam, TCharacteristics>
+{
+    5: required QsTypeKind Kind = InvalidType;
+
+    // Not null when Kind is ArrayType.
+    10: nullable<TData> ArrayType = nothing;
+
+    // Not null when Kind is TupleType.
+    15: nullable<vector<TData>> TupleType = nothing;
+
+    // Not null when Kind is UserDefinedType.
+    20: nullable<TUdt> UserDefinedType = nothing;
+
+    // Not null when Kind is TypeParameter.
+    25: nullable<TParam> TypeParameter = nothing;
+
+    // Not null when Kind is Operation.
+    30: nullable<QsTypeKindOperation<TData, TCharacteristics>> Operation = nothing;
+
+    // Not null when Kind is Function.
+    35: nullable<QsTypeKindFunction<TData>> Function = nothing;
+}
+
+struct ResolvedType
+{
+    5: required QsTypeKindComposition<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation> TypeKind;
+}
+
+struct InferredExpressionInformation
+{
+    5: required bool IsMutable;
+
+    10: required bool HasLocalQuantumDependency;
+}
+
+struct TypedArgument
+{
+    5: required QsQualifiedName Callable;
+
+    10: required string Name;
```

Comment:
You are making frequent use of `required` in a way that I think will not beneficial. In fact, the [Schema Evolution](https://microsoft.github.io/bond/manual/bond_cs.html#schema-evolution) section of the bond guide lists this as a best practice: `required` should be used sparingly and only with careful consideration.

The `required` attribute has two major effects: during deserialization if that field is missing it is treated as a deserialization error, and during serialization it will never be omitted when it is the default value. By contrast, bond fields are optional by default, meaning their absence during deserialization is not a failure and they may be omitted during serialization if they are equal to their default. The biggest place this makes a difference is with compatibility across versions, which is likely less of a concern here, but still with thinking through. I'm not certain this will help or hurt performance, so take this all as me thinking out loud and not necessarily as a recommendation.

Consider this case, a `string Name` property. If it is required, you are saying that the field must be present in order to pass deserialization. But an empty string is still a valid string, since that is the default value. So imagining a struct with one field that is `required string Name;` and deserializing from JSON, this would be considered invalid:
```json
{ }
```
But both of these are actually valid and would deserialize successfully:
```json
{ "Name": "foo" }
```
and 
```json
{ "Name": "" }
```
So `required` doesn't actually provide any additional protection against the empty string; you'd still need to check for it and reject it after deserializing the data. To match this, if your data structure is being serialized and has `Name` equal to empty string, it will serialize into `{ "Name": "" }`.

 If instead the `Name` property were not required, both `{ "Name": "" }` and `{ }` would deserialize into an object with Name equal to the empty string (which your code already has to handle), and would always serialize into `{ }`, saving space in the generated encoding. This tends to be better for both serialized data size and future compatibility.

Sorry, I know that was a long explanation, but hopefully that helps provide context to decide whether all these fields really need to be required or whether most or all of them can be switched to optional.

Reply:
Yes, this helps a lot. I will go back and be more careful about using required.

Reply:
Sounds good! Let me know if there are any specific fields you’d like to discuss the trade-offs for. Reviewing the guidance is making me reevaluate the couple places where I’ve used it, so now I’m going back through those and rethinking them too. There is also a `required_optional` that splits the difference and is used for compatibility to allow gradually converting between the two.

Reply:
I ended up removing the `required` from almost all fields. The only ones I left as `required` were the ones that are composite structs in which depending on the kind a different field is the one that contains specific data. I did this because if the kind is really required in those cases.

Thanks for the review! I think the schema is now much more flexible and it aligns with our purpose of making the data structures more efficient without breaking compatibility.